Shader optimizations to reduce VGPR usage and increase occupancy

reduz opened the pull request:
Reduced from 116 to 80 VGPRs in normal usage, still more work is needed.

Some not often used features are gone for the sake of this and some are disabled, will re-enable later.
-Shadow color per light is gone
-Oren Nayar is gone (implement manually if you really want it, makes little difference)
-SSAO color is gone
-SSAO ao affect is gone because this confused users, min(ao,ssao) now used, should work in most cases.
Also

-softshadows (based on angular distance) are likely going to be removed from directional light, too expensive dont look great, and dont make much of a difference
-projectors are disabled.

reduz committed:
Shader optimizations to reduce VGPR usage and increase occupancy

reduz commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -6288,13 +6283,14 @@ void RendererSceneRenderRD::_setup_lights(const PagedArray<RID> &p_lights, const
 
-				light_data.attenuation_energy[0] = Math::make_half_float(storage->light_get_param(base, RS::LIGHT_PARAM_ATTENUATION));
-				light_data.attenuation_energy[1] = Math::make_half_float(sign * storage->light_get_param(base, RS::LIGHT_PARAM_ENERGY) * Math_PI);
+				light_data.attenuation = storage->light_get_param(base, RS::LIGHT_PARAM_ATTENUATION);
```
This just cleaned up how parameters are sent to the reflection probe struct

reduz commented on servers/rendering/renderer_rd/renderer_scene_render_rd.h:
```diff
@@ -1313,10 +1322,15 @@ class RendererSceneRenderRD : public RendererSceneRender {
-			uint8_t color_specular[4]; //rgb color, a specular (8 bit unorm)
-			uint16_t cone_attenuation_angle[2]; // attenuation and angle, (16bit float)
-			uint8_t shadow_color_enabled[4]; //shadow rgb color, a>0.5 enabled (8bit unorm)
+
```
Likewise cleaned up lights

clayjohn commented on servers/rendering/renderer_rd/shaders/scene_forward.glsl:
```diff
@@ -591,14 +598,25 @@ LIGHT_SHADER_CODE
 	if (metallic < 1.0) {
+		float roughness = unpackUnorm4x8(orms).y;
+
 #if defined(DIFFUSE_OREN_NAYAR)
```
If deleting Oren-Nayar, it should be deleted here and in the material and shader as well

Chaosus replied:
Fixed by https://github.com/godotengine/godot/pull/49419

clayjohn commented on servers/rendering/renderer_rd/shaders/scene_forward.glsl:
```diff
@@ -2710,173 +3009,149 @@ FRAGMENT_SHADER_CODE
-	}
-#endif //LOW_END_MODE
+	// multiply by albedo
+	diffuse_light *= albedo; // ambient must be multiplied by albedo at the end
```
comment out of date

clayjohn commented on servers/rendering/renderer_rd/shaders/scene_forward.glsl:
```diff
@@ -2297,6 +2335,14 @@ FRAGMENT_SHADER_CODE
+#ifndef LOW_END_MODE
+	if (scene_data.ssao_enabled) {
+		float ssao = texture(sampler2D(ao_buffer, material_samplers[SAMPLER_LINEAR_CLAMP]), screen_uv).r;
+		ao = min(ao, ssao);
```
We need to delete SSAO_ao_affect from environment etc.

Calinou replied:
For reference, I opened an issue to track this: https://github.com/godotengine/godot/issues/54611

clayjohn reviewed (commented):
Looks mostly good!

reduz committed:
Added GPU based cluster builder

Clustering is now GPU based, uses an implementation based on the Activision algorithm.

clayjohn commented:
Im getting 2 types of artifacts with shaders. One is random blocks of shadow that flicker in and out (likely entire clusters missing a light). And the second is that the derivative shadow optimization doesnt scale well with distance from a shadow. So when you get far enough away from a shadow the edges take on a very aliased look

![Screenshot (60)](https://user-images.githubusercontent.com/16521339/104856715-2eb1b100-58c9-11eb-8943-8714adc86f8c.png)

clayjohn commented on servers/rendering/renderer_rd/shaders/cluster_render.glsl:
```diff
@@ -0,0 +1,168 @@
+			// threads that did break; go inactive and no longer count
+			uint first = subgroupBroadcastFirst(cluster_offset);
+			// update the mask for thread that match this cluster
+			uvec4 mask = subgroupBallot(first == cluster_offset);
```
```suggestion
			mask = subgroupBallot(first == cluster_offset);
```

This line is causing the issue that I posted about.

reduz replied:
you are a genius, thanks! fixed.

reduz commented:
@akien-mga Should hopefully be ok to merge but the static checks are not passing and I already have clang-format-11.

Calinou commented:
@reduz I would recommend avoiding nested ternary operators, it should fix the clang-format issue at the same time.

clayjohn commented:
I haven't reviewed the code yet, but running this build there are some artifacts created on triangle edges. It must be due to the shadow derivative path.

![Screenshot (72)](https://user-images.githubusercontent.com/16521339/104936378-4a689600-5961-11eb-812e-cdaac9d32dcf.png)

reduz commented:
@clayjohn disabled it for now

JFonS commented:
I went over the files, it wasn't an in-depth review but I didn't spot anything wrong.

Other than that, Github lists some files as edited, but they don't contain any changes, not sure what's going on. Also, static checks have failed, I think they should be restarted.

akien-mga commented:
I restarted CI jobs for static checks and Android which failed for a seemingly unrelated reason, likely a CI oops.

akien-mga commented:
> I restarted CI jobs for static checks and Android which failed for a seemingly unrelated reason, likely a CI oops.

Ah no it's not a CI issue, @reduz did remove executable permission on those files, so they can't run on CI:
![Screenshot_20210119_231914](https://user-images.githubusercontent.com/4701338/105099953-d0452800-5aac-11eb-97a9-baf5d7a879b4.png)

akien-mga commented:
Thanks!
